### PR TITLE
8346394: Bundled freetype library needs to have JNI_OnLoad for static builds

### PR DESCRIPTION
**Pull request by magicus**
All JDK native libraries needs to have a proper `JNI_OnLoad`, set either by `DEF_STATIC_JNI_OnLoad` or `DEF_JNI_OnLoad`. This is so that in a static build, the JVM can recognize that the library is already loaded, and does not try to load a (non-existent) external library.

Our bundled version of freetype is missing such a marker.

This turns out to be a problem just on Windows, since the `System.loadLibrary("freetype")` code in `FontManagerNativeLibrary` is only executed on Windows. However, the cost of including this in the bundled versions on all platforms is very small, and it is a safeguard for any future changes. In practice, freetype bundling is only enabled on Windows anyways.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8346394](https://bugs.openjdk.org/browse/JDK-8346394): Bundled freetype library needs to have JNI_OnLoad for static builds (**Bug** - P3)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**) Review applies to [13c93e63](https://git.openjdk.org/jdk/pull/22791/files/13c93e6376c72f15882e22cde315decf662f5d8e)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22791/head:pull/22791` \
`$ git checkout pull/22791`

Update a local copy of the PR: \
`$ git checkout pull/22791` \
`$ git pull https://git.openjdk.org/jdk.git pull/22791/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22791`

View PR using the GUI difftool: \
`$ git pr show -t 22791`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22791.diff">https://git.openjdk.org/jdk/pull/22791.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22791#issuecomment-2548398199)
</details>

### CODE REVIEW

**Review by erikj79 — APPROVED** [review]

**Comment by prrace** [comment]
If I understand correctly, this is a "new" requirement because of the need to support static builds ?

Seems odd if the library doesn't actually use JNI. And what if the JDK is trying to load  a library that is not part of the JDK ?

>  freetype bundling is only enabled on Windows anyways.

We bundle it on mac too ..

**Comment by prrace** [comment]
> > We bundle it on mac too ..
> 
> Oh, that's right. But the macOS bundling is done purely at link time, so we do not have this problem, as far as I can tell.

I see (I think)

File: src/java.desktop/share/native/libfreetype/java_freetype.c [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "jni_util.h"
+
```
Comment (prrace) [inline src/java.desktop/share/native/libfreetype/java_freetype.c:27]: A comment here explaining why this is needed might be a good idea.

**Comment by dholmes-ora** [comment]
A bit late but why do we presume that just because this library is bundled that it has to be part of the static image?

This static build stuff continues to leave me with a "bad taste in my mouth".

**Comment by magicus** [comment]
> A bit late but why do we presume that just because this library is bundled that it has to be part of the static image?

No? The library can be bundled as a dynamic library as well, that has always been and still is the case.

> This static build stuff continues to leave me with a "bad taste in my mouth".

I agree that the static build stuff is still not up to normal JDK standards. It was originally "tacked on" from the side, with a lot of kludges to get it to work. I am trying to work through these problematic cases and clean up the code. 

This current kludge is that there is a lot of code doing System.loadLibrary(someInternalLibrary), and for static builds these calls should really be a no-op. But how do you know? The solution chosen by the original implementors of static builds was to look for a symbol JNI_OnLoad_someInternalLibrary, and if it was found, skip loading the dynamic library. It can certainly be discussed if this was a good way of doing things, but that is how it currently works. (It does have the advantage that users can add static libraries of their own in this way.)

**Comment by dholmes-ora** [comment]
> The solution chosen by the original implementors of static builds was to look for a symbol JNI_OnLoad_someInternalLibrary, and if it was found, skip loading the dynamic library. 

Is that necessary? Presumably if you try to dynamically load a statically linked library it will fail. Does the failure indicate the reason? If so then just ignore it.

> It does have the advantage that users can add static libraries of their own in this way.

Is this intended to be a supported mechanism? That needs to be specified and documented somewhere if so.

**Comment by magicus** [comment]
> > The solution chosen by the original implementors of static builds was to look for a symbol JNI_OnLoad_someInternalLibrary, and if it was found, skip loading the dynamic library.
> 
> Is that necessary? Presumably if you try to dynamically load a statically linked library it will fail. Does the failure indicate the reason? If so then just ignore it.

It is perhaps not necessary, but this is the way it is implemented, and had been for years. Just as you like to be conservative and don't change things that's been working for years, so am I a bit reluctant to change this behavior. Doing so would be a massive undertaking, on a completely different scale than just fixing a few issues to get the Windows static launcher to work properly.

> > It does have the advantage that users can add static libraries of their own in this way.
> 
> Is this intended to be a supported mechanism? That needs to be specified and documented somewhere if so.

Yes it is. It is documented in the JNI specification: https://docs.oracle.com/en/java/javase/23/docs/specs/jni/invocation.html#jni_onload_l

**Pull request closed without merging**